### ISSUE TITLE
Splitting source property name to destination property's sub-properties

### ISSUE THREAD

**Issue by satano** [MapsterMapper/Mapster#163]
Mapster is able to flatten source object property to destination property ([Mappable objects in wiki](https://github.com/MapsterMapper/Mapster/wiki/Data-types#mappable-objects)) but not vice versa.

I have following classes.

``` csharp
public class InvoiceDto
{
    public int Id { get; set; }
    public string InvoiceNumber { get; set; }
    public string SupplierName { get; set; }
    public string SupplierAddress { get; set; }
}


public class ContractingParty
{
    public string Name { get; set; }
    public string Address { get; set; }
}

public class Invoice
{
    public int Id { get; set; }
    public string InvoiceNumber { get; set; }
    public ContractingParty Supplier { get; } = new ContractingParty();
}
```

Now I can map `Invoice` instance to `InvoiceDto`, but not the other way around.

``` csharp
var invoice = new Invoice();
invoice.Id = 1;
invoice.InvoiceNumber = "123456";
invoice.Supplier.Name = "Test supplier";
invoice.Supplier.Address = "The big city";

// This works - adaptedDto.SupplierName is set to "Test supplier"
var adaptedDto = invoice.Adapt<InvoiceDto>();

var invoiceDto = new InvoiceDto();
invoiceDto.Id = 1;
invoiceDto.InvoiceNumber = "123456";
invoiceDto.SupplierName = "Test supplier";
invoiceDto.SupplierAddress = "The big city";

// This does not work - adaptedInvoice.Supplier.Name is not set.
var adaptedInvoice = invoiceDto.Adapt<Invoice>();
```

It is clear from the documentation that it works only one way. It can flatten source nested properties, but not "split" property to nested properties. To make it work, we have `AfterMapping` configured:

``` csharp
TypeAdapterConfig<InvoiceDto, Invoice>
    .NewConfig()
    .AfterMapping((src, dest) =>
    {
        dest.Supplier.Name = src.SupplierName;
        dest.Supplier.Address = src.SupplierAddress;
    });
```

It works, but there is a drawback. If a new property is introduced to `ContractingParty` class, it must be added to mapping. This can be easily forgotten. And of course, the number of properties is much bigger.

So I have two questions.

1. Is there some other (simpler) way how to configure mapping to achieve this?
2. Why is this not working the other way? Is there som performance issue, or maybe it is not so easy to implement?

**Comment by chaowlert** [comment]
Hi @satano,

I also would like to have unflattening feature. For my case is to deal with flat file upload mapping to object.

To answer your questions
1. What I did is same as yours, I mapped manually. No magic.
2. I believe solution for this problem is to support 2-ways mapping. Where we already got mapping pairs between source and destination. Then we just reverse the mapping.

**Comment by satano** [comment]
Thanks @chaowlert 

Do you have any internal plan implementing this?

**Comment by chaowlert** [comment]
I would love to improve Mapster. I also would like to apply Roslyn for code generation & debugging.
But I don't have spare time at any time soon. So please don't wait.

**Comment by satano** [comment]
Sure. Thanks for the answers.

**Comment by chaowlert** [comment]
implemented in 4.0
